pkg/policy/directory/watcher.go

```diff
@@ -0,0 +1,222 @@
+	}
+
+	ns := cnp.ObjectMeta.Namespace
+	// For clusterwide policy namespace will be empty.
+	if ns != "" {
+		nsLabel := labels.NewLabel(k8sConst.PolicyLabelNamespace, ns, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, nsLabel)
+		srcLabel := labels.NewLabel("policy-derived-from", k8sCiliumUtils.ResourceTypeCiliumNetworkPolicy, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, srcLabel)
+	} else {
+		srcLabel := labels.NewLabel("policy-derived-from", k8sCiliumUtils.ResourceTypeCiliumClusterwideNetworkPolicy, labels.LabelSourceDirectory)
+		labelsArr = append(labelsArr, srcLabel)
+	}
+
+	return labelsArr
+}
+
+// read cilium network policy yaml file and convert to policy object and
+// add rules to policy engine.
+func (p *policyWatcher) addToPolicyEngine(cnpFilePath string) error {
+	// read from file and convert to cnp object
+	cnp, err := p.translateToCNPObject(cnpFilePath)
+	if err != nil {
+		return err
+	}
+
+	fileName := filepath.Base(cnpFilePath)
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		p.config.StaticCNPPath,
+		fileName,
+	)
+
+	// convert to rules
+	rules, err := cnp.Parse()
+	if err != nil {
+		return err
+	}
+
+	// update labels
+	lbls := getLabels(fileName, cnp)
+	for _, r := range rules {
+		r.Labels = lbls
+	}
+
+	// add to policy engine
+	_, err = p.policyManager.PolicyAdd(rules, &policy.AddOptions{
+		ReplaceByResource: true,
+		Source:            source.Directory,
+		Resource:          resourceID,
+	})
+
+	if err == nil {
+		p.fileNameToCnpCache[fileName] = cnp
+	}
+
+	return err
+}
+
+func (p *policyWatcher) deleteFromPolicyEngine(cnpFilePath string) error {
+	fileName := filepath.Base(cnpFilePath)
+	cnp := p.fileNameToCnpCache[fileName]
+	if cnp == nil {
+		p.log.WithField("file", fileName).Error("BUG: Policy deletion request for file which was never added")
+		return nil
+	}
+
+	resourceID := ipcacheTypes.NewResourceID(
+		ipcacheTypes.ResourceKindFile,
+		p.config.StaticCNPPath,
+		fileName,
+	)
+	_, err := p.policyManager.PolicyDelete(getLabels(fileName, cnp), &policy.DeleteOptions{
+		Source:           source.Directory,
+		DeleteByResource: true,
+		Resource:         resourceID})
+
+	delete(p.fileNameToCnpCache, fileName)
+	return err
+}
+
+func (p *policyWatcher) watchDirectory(ctx context.Context) {
+	go func() {
+		p.log.Info("Directory policy watcher started")
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			p.log.WithError(err).Error("Initializing NewWatcher failed")
+			return
+		}
+		defer watcher.Close()
+
+		dir := p.config.StaticCNPPath
+		err = watcher.Add(dir)
+		if err != nil {
+			p.log.WithError(err).WithField(logfields.Path, dir).Error("Failed to watch policy directory. Policies will not be loaded from disk")
+			return
+		}
+
+		files, err := os.ReadDir(dir)
+		if err != nil {
+			p.log.WithError(err).WithField("Directory", dir).Warn("Failed to read policy directory")
```

Comment:
Let's make this an error, since it's pretty fatal. Also, preferably use lower-case field names:

```suggestion
			p.log.WithError(err).WithField(logfields.Path, dir).Error("Failed to read policy directory")
```